src/npc.cpp

```diff
@@ -1857,7 +1860,9 @@ void npc::on_attacked( const Creature &attacker )
     }
     if( attacker.is_avatar() && !is_enemy() && !is_dead() ) {
         make_angry();
-        hit_by_player = true;
+        if( !guaranteed_hostile() ) {
+            hit_by_player = true;
```

Comment:
This causes `badguy` assignment to always evaluate true for non-one-hit kills as well, since the logic is:
`badguy = ( guaranteed_hostile() || !hit_by_player )`
`hit_by_player = !guaranteed_hostile() && on_attacked`